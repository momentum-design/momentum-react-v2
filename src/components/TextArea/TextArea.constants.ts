import {
  AUTO_CAPITALIZE,
  AUTO_COMPLETE,
} from '@momentum-design/components/dist/components/input/input.constants.js';
import { POPOVER_PLACEMENT } from '@momentum-design/components/dist/components/popover/popover.constants.js';
import { VALIDATION } from '@momentum-design/components/dist/components/formfieldwrapper/formfieldwrapper.constants.js';
import { WRAP } from '@momentum-design/components/dist/components/textarea/textarea.constants.js';

const CLASS_PREFIX = 'md-text-area';

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { AUTO_CAPITALIZE, AUTO_COMPLETE, POPOVER_PLACEMENT, VALIDATION, WRAP, STYLE };
