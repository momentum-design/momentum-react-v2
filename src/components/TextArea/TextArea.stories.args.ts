import { commonStyles } from '../../storybook/helper.stories.argtypes';
import {
  AUTO_CAPITALIZE,
  AUTO_COMPLETE,
  POPOVER_PLACEMENT,
  VALIDATION,
  WRAP,
} from './TextArea.constants';

export default {
  ...commonStyles,
  name: {
    description:
      'Indicates the name of the component group. They are used to group elements in a form together.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: '',
      },
    },
  },
  label: {
    description:
      'The label of the input field. It is linked to the input field using the for attribute.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: '',
      },
    },
  },
  rows: {
    description: 'The rows attribute specifies the visible number of lines in a text area.',
    control: { type: 'number' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: 5,
      },
    },
  },
  cols: {
    description: 'The cols attribute specifies the visible number of characters in a text area.',
    control: { type: 'number' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: 40,
      },
    },
  },
  wrap: {
    description:
      'The wrap attribute specifies how the text in a text area is to be wrapped when submitted in a form.',
    options: [...Object.values(WRAP)],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'soft',
      },
    },
  },
  required: {
    description:
      'The required attribute to indicate that the input field is required. It is used to append a required indicator (*) to the label.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: false,
      },
    },
  },
  placeholder: {
    description: 'The placeholder text that is displayed when the textarea field is empty.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: '',
      },
    },
  },
  value: {
    description:
      'Indicates the value of the component group (ex: input, checkbox, radio, select etc...)',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: '',
      },
    },
  },
  helpText: {
    description: 'The help text that is displayed below the input field.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: '',
      },
    },
  },
  helpTextType: {
    description:
      "The type of help text. It can be 'default', 'error', 'warning', 'success', 'priority'.",
    options: [...Object.values(VALIDATION)],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: VALIDATION.DEFAULT,
      },
    },
  },
  readonly: {
    description:
      'readonly attribute of the textarea field. If true, the textarea field is read-only.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: false,
      },
    },
  },
  disabled: {
    description:
      'Indicates whether the component is disabled. When the component is disabled for user interaction; it is not focusable or clickable.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: false,
      },
    },
  },
  autoCapitalize: {
    description: 'The autocapitalize attribute of the textarea field.',
    options: [...Object.values(AUTO_CAPITALIZE)],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'AutoCapitalizeType',
      },
      defaultValue: {
        summary: 'off',
      },
    },
  },
  autoComplete: {
    description: 'The autocomplete attribute of the textarea field.',
    options: [...Object.values(AUTO_COMPLETE)],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'AutoCompleteType',
      },
      defaultValue: {
        summary: 'off',
      },
    },
  },
  autoFocus: {
    description: 'If true, the textarea field is focused when the component is rendered.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: false,
      },
    },
  },
  ariaLabel: {
    description:
      'Defines a string value that labels the current element. The Aria-Label attribute to be set for accessibility.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: '',
      },
    },
  },
  dirName: {
    description:
      'Specifies the name of the directionality of text for submission purposes (e.g., "rtl" for right-to-left).',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: '',
      },
    },
  },
  maxLength: {
    description: 'The maximum number of characters that the textarea field can accept.',
    control: { type: 'number' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  minLength: {
    description: 'The minimum number of characters that the textarea field can accept.',
    control: { type: 'number' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  maxCharacterLimit: {
    description: 'maximum character limit for the textarea field for character counter.',
    control: { type: 'number' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  validationMessage: {
    description:
      'Custom validation message that will override the default message and displayed when the input is invalid.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  toggletipText: {
    description:
      'The toggletip text that is displayed when the label is hovered. It is used to provide additional information about the label.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  toggletipPlacement: {
    description: 'The placement of the toggletip that is displayed when the info icon is hovered.',
    options: [...Object.values(POPOVER_PLACEMENT)],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'top',
      },
    },
  },
  infoIconArialLabel: {
    description:
      'Aria label for the info icon that is displayed next to the label when toggletipText is set. This is used for accessibility purposes to provide a description of the icon.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },

  onLimitExceeded: {
    action: 'onLimitExceeded',
    description: 'This event is dispatched once when the character limit exceeds or restored.',
    table: {
      category: 'Events',
      type: {
        summary: '(e: Event) => void',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  onInput: {
    action: 'onInput',
    description:
      'This event is dispatched when the value of the textarea field changes (every press).',
    table: {
      category: 'Events',
      type: {
        summary: '(e: InputEvent) => void',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  onChange: {
    action: 'onChange',
    description:
      ' This event is dispatched when the value of the textarea field changes (on blur).',
    table: {
      category: 'Events',
      type: {
        summary: '(e: Event) => void',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  onFocus: {
    action: 'onFocus',
    description: ' This event is dispatched when the textarea receives focus.',
    table: {
      category: 'Events',
      type: {
        summary: '(e: FocusEvent) => void',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  onBlur: {
    action: 'onBlur',
    description: ' This event is dispatched when the textarea loses focus.',
    table: {
      category: 'Events',
      type: {
        summary: '(e: FocusEvent) => void',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
};
