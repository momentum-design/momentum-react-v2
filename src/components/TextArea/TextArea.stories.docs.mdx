The `<TextArea />` component. This component is used is used to get the multi-line text input from the user.

Rendering TextArea from momentum-design (see more details here: https://momentum-design.github.io/momentum-design/storybook-static/index.html?path=/story/components-textarea--example)