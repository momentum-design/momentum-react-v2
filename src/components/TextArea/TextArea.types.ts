import type { MdcTextareaProps } from '../../types';

export type Props = MdcTextareaProps;
