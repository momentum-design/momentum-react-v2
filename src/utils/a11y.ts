import { AriaAttributes, useEffect } from 'react';
import { RequireOneOf } from './types';

export type AriaLabelRequired = RequireOneOf<AriaAttributes, ['aria-label', 'aria-labelledby']>;

/**
 * Check if `aria-label` or `aria-labelledby` are defined and are truthy
 *
 * @param componentName The name of the component
 * @param props The object of props passed to the component
 */
export function useCheckAriaLabel(
  componentName: string,
  props: { 'aria-label'?: string; 'aria-labelledby'?: string }
): void {
  useEffect(() => {
    if (!props['aria-label'] && !props['aria-labelledby']) {
      console.warn(`MRV2: ${componentName} requires aria-label or aria-labelledby.`);
    }
  }, [componentName, props]);
}
