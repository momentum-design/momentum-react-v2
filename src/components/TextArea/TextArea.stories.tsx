import React, { useState } from 'react';

import { Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import { TextAreaProps } from '.';
import TextArea from './TextArea';
import Documentation from './TextArea.stories.docs.mdx';
import argTypes from './TextArea.stories.args';
import ButtonPill from '../ButtonPill';

export default {
  title: 'Momentum Ui/TextArea',
  component: TextArea,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
  args: {
    label: 'Label',
    helpText: 'Help text',
    helpTextType: 'default',
    placeholder: 'Placeholder',
  },
};

const Example = Template<TextAreaProps>((props: TextAreaProps) => {
  const [inputValue, setInputValue] = useState('');
  const handleInput = (e: InputEvent) => {
    const value = (e.target as HTMLTextAreaElement).value;
    setInputValue(value);
    // eslint-disable-next-line no-console
    console.log('example inputvalue', value);
  };

  return <TextArea value={inputValue} required {...props} onInput={handleInput} />;
}).bind({});
Example.argTypes = { ...argTypes };

const variants: Array<{
  type: TextAreaProps['helpTextType'];
  label: string;
  helpText: string;
}> = [
  { type: 'default', label: 'Default Label', helpText: 'Default help text' },
  { type: 'error', label: 'Error Label', helpText: 'Error help text' },
  { type: 'warning', label: 'Warning Label', helpText: 'Warning help text' },
  { type: 'success', label: 'Success Label', helpText: 'Success help text' },
  { type: 'priority', label: 'Priority Label', helpText: 'Priority help text' },
];
const AllVariants = Template<TextAreaProps>((props: TextAreaProps) => {
  return (
    <div
      style={{
        display: 'grid',
        gridTemplateColumns: `repeat(3, auto)`,
        gap: '1rem',
        alignItems: 'end',
      }}
    >
      {variants.map(({ type, label, helpText }) => (
        <TextArea key={type} {...props} helpTextType={type} label={label} helpText={helpText} />
      ))}
    </div>
  );
}).bind({});
AllVariants.argTypes = { ...argTypes };

const Disabled = Template<TextAreaProps>((props: TextAreaProps) => {
  return (
    <div>
      <TextArea value="Disabled text area" disabled {...props} />
    </div>
  );
}).bind({});
Disabled.argTypes = { ...argTypes };

const Readonly = Template<TextAreaProps>((props: TextAreaProps) => {
  return (
    <div>
      <TextArea value="Readonly text area" readonly {...props} />
    </div>
  );
}).bind({});
Readonly.argTypes = { ...argTypes };

const Form = Template<TextAreaProps>((props: TextAreaProps) => {
  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const formData = new FormData(event.currentTarget);
    const selectedValue = formData.get('formTextarea');
    // eslint-disable-next-line no-console
    console.log('Form submitted value:', selectedValue);
  };

  return (
    <form onSubmit={handleSubmit}>
      <TextArea
        name="formTextarea"
        id="textarea"
        required
        toggletipText="toggletip"
        maxCharacterLimit={100}
        {...props}
      />
      <div style={{ display: 'flex', gap: '1.5rem', marginTop: '1rem' }}>
        <ButtonPill type="submit" size={28} style={{ width: '5rem' }}>
          Submit
        </ButtonPill>
        <ButtonPill type="reset" size={28} style={{ width: '5rem' }}>
          Reset
        </ButtonPill>
      </div>
    </form>
  );
}).bind({});
Form.argTypes = { ...argTypes };

export { Example, AllVariants, Disabled, Readonly, Form };
