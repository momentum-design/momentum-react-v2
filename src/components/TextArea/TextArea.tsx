import React, { forwardRef, RefObject, useRef } from 'react';
import { Textarea as MdcTextarea } from '@momentum-design/components/dist/react';
import type { Textarea } from '@momentum-design/components';
import classnames from 'classnames';

import { STYLE } from './TextArea.constants';
import { Props } from './TextArea.types';

const TextArea = forwardRef((props: Props, providedRef: RefObject<Textarea>) => {
  const { className, id, style, ...rest } = props;
  const internalRef = useRef();
  const ref = providedRef || internalRef;

  return (
    <MdcTextarea
      ref={ref}
      className={classnames(className, STYLE.wrapper)}
      id={id}
      style={style}
      {...rest}
    />
  );
});

TextArea.displayName = 'TextArea';
export default TextArea;
