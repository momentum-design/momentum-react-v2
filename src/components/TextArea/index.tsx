import { default as TextArea } from './TextArea';
import { Props } from './TextArea.types';
import * as CONSTANTS from './TextArea.constants';

export type TextAreaProps = Props;
export { CONSTANTS as TEXT_AREA_CONSTANTS };

export default TextArea;
