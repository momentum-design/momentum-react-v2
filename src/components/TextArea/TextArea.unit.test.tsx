import React from 'react';
import { renderWithWebComponent } from '../../../test/utils';
import TextArea, { TEXT_AREA_CONSTANTS as CONSTANTS } from './';

describe('<TextArea />', () => {
  const setup = async (component: React.ReactElement) => {
    return renderWithWebComponent(component);
  };

  describe('snapshot', () => {
    it('should match snapshot', async () => {
      expect.assertions(1);

      const { container } = await setup(<TextArea />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', async () => {
      expect.assertions(1);

      const className = 'example-class';
      const { container } = await setup(<TextArea className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', async () => {
      expect.assertions(1);

      const id = 'example-id';
      const { container } = await setup(<TextArea id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', async () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const { container } = await setup(<TextArea style={style} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its wrapper class', async () => {
      expect.assertions(1);

      const { container } = await setup(<TextArea />);
      const element = container.querySelector('mdc-textarea');

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', async () => {
      expect.assertions(1);

      const className = 'example-class';

      const { container } = await setup(<TextArea className={className} />);
      const element = container.querySelector('mdc-textarea');

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', async () => {
      expect.assertions(1);

      const id = 'example-id';

      const { container } = await setup(<TextArea id={id} />);
      const element = container.querySelector('mdc-textarea');

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', async () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const { container } = await setup(<TextArea style={style} />);
      const element = container.querySelector('mdc-textarea');

      expect(element.getAttribute('style')).toBe(styleString);
    });
  });
});
