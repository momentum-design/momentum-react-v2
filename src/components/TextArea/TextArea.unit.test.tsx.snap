// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> snapshot should match snapshot 1`] = `
<div>
  <mdc-textarea
    aria-invalid="false"
    class="md-text-area-wrapper"
    help-text-type="default"
    internals-valid=""
    name=""
    toggletip-placement="top"
    value=""
  />
</div>
`;

exports[`<TextArea /> snapshot should match snapshot with className 1`] = `
<div>
  <mdc-textarea
    aria-invalid="false"
    class="example-class md-text-area-wrapper"
    help-text-type="default"
    internals-valid=""
    name=""
    toggletip-placement="top"
    value=""
  />
</div>
`;

exports[`<TextArea /> snapshot should match snapshot with id 1`] = `
<div>
  <mdc-textarea
    aria-invalid="false"
    class="md-text-area-wrapper"
    help-text-type="default"
    internals-valid=""
    name=""
    toggletip-placement="top"
    value=""
  />
</div>
`;

exports[`<TextArea /> snapshot should match snapshot with style 1`] = `
<div>
  <mdc-textarea
    aria-invalid="false"
    class="md-text-area-wrapper"
    help-text-type="default"
    internals-valid=""
    name=""
    style="color: pink;"
    toggletip-placement="top"
    value=""
  />
</div>
`;
